store/prefixstore.go

```diff
@@ -13,6 +14,18 @@ type prefixStore struct {
 	prefix []byte
 }
 
+func clone(bz []byte) (res []byte) {
+	res = make([]byte, len(bz))
+	copy(res, bz)
+	return
+}
+
+func (s prefixStore) key(key []byte) (res []byte) {
+	res = clone(s.prefix)
+	res = append(res, key...)
```

Comment:
Actually, I had a problem using `s.prefix` without copying. Since `[]byte` is a reference type and we pass it to the iavl tree(as `key`), each time we call `prefixStore.(Get/Set/Has...)` the `[]byte` stored in the tree is modified. Also see #1912. 

So I think cloning `s.prefix` is necessary but we can still allocate enough space for `clone`ing when we are `append`ing after that.

Reply:
The issue is if the slice capacity of the prefix is larger than then length of the prefix  and you assign a key that fits within the prefix's capacity.

For example if prefix has length 6 and capacity 9 the backing array looks like:
```
prefix = |prefix|---|
// then call k1 := Key(abc) 
k1 = |prefixabc|
prefix = |prefix|abc|
// then call k2 := Key(cde)
k2 = |prefixcde|
prefix = |prefix|cde|
```

If you reslice with capacity `[start:end:cap`] you will avoid this from happening though, which was why that syntax was introduced... Or if you initialise prefix from a string. You'll always get a copy.

This looks cleaner to me, but I don't know if there is any performance benefit versus a `make` with the correct concatenated length - ultimately the same copy will happen.